vault/ttl.go

```diff
@@ -0,0 +1,90 @@
+package vault
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/hashicorp/vault/logical"
+)
+
+func calculateTTL(sysView logical.SystemView, increment, backendTTL, period, backendMaxTTL, explicitMaxTTL time.Duration, startTime time.Time) (ttl time.Duration, warnings []string, errors error) {
+	// Truncate all times to the second since that is the lowest precision for
+	// TTLs
+	now := time.Now().Truncate(time.Second)
+	if startTime.IsZero() {
+		startTime = now
+	} else {
+		startTime = startTime.Truncate(time.Second)
+	}
+
+	// Start off with the sys default value, and update according to period/TTL
+	// from resp.Auth
+	ttl = sysView.DefaultLeaseTTL()
+
+	// Use the mount's configured max unless the backend specifies
+	// something more restrictive (perhaps from a role configuration
+	// parameter)
+	maxTTL := sysView.MaxLeaseTTL()
+	if backendMaxTTL > 0 && backendMaxTTL < maxTTL {
+		maxTTL = backendMaxTTL
+	}
+	if explicitMaxTTL > 0 && explicitMaxTTL < maxTTL {
+		maxTTL = explicitMaxTTL
+	}
+
+	// Should never happen, but guard anyways
+	if maxTTL < 0 {
```

Comment:
I believe it'd also be an error if max ttl was 0 (same for ttl above) since we should always fall back to system level defaults.